python/dgl/mock_sparse/diag_matrix.py

```diff
@@ -0,0 +1,229 @@
+        """
+        return self.val.device
+
+    def as_sparse(self) -> SparseMatrix:
+        """Convert the diagonal matrix into a sparse matrix object
+
+        Returns
+        -------
+        SparseMatrix
+            The converted sparse matrix object
+
+        Example
+        -------
+
+        >>> import torch
+        >>> val = torch.randn(5)
+        >>> mat = diag(val)
+        >>> sp_mat = mat.as_sparse()
+        """
+        row = col = torch.arange(len(self.val)).to(self.device)
+        return create_from_coo(row, col, self.val, self.shape)
+
+def diag(val: torch.Tensor, shape: Optional[Tuple[int, int]] = None) -> DiagMatrix:
+    """Create a diagonal matrix based on the diagonal values
+
+    Parameters
+    ----------
+    val : torch.Tensor
+        Diagonal of the matrix. It can take shape (N) or (N, D).
+    shape : tuple[int, int], optional
+        If not specified, it will be inferred from :attr:`val`, i.e.,
+        (N, N). Otherwise, :attr:`len(val)` must be equal to :attr:`min(shape)`.
+
+    Returns
+    -------
+    DiagMatrix
+        Diagonal matrix
+
+    Examples
+    --------
+
+    Case1: 5-by-5 diagonal matrix with scaler values on the diagonal
+
+    >>> import torch
+    >>> val = torch.randn(5)
+    >>> mat = diag(val)
+    >>> mat.shape
+    (5, 5)
+    >>> mat.nnz
+    5
+
+    Case2: 5-by-10 diagonal matrix with scaler values on the diagonal
+
+    >>> val = torch.randn(5)
+    >>> mat = diag(val, shape=(5, 10))
+    >>> mat.shape
+    (5, 10)
+    >>> mat.nnz
+    5
+
+    Case3: 5-by-5 diagonal matrix with tensor values on the diagonal
+
+    >>> val = torch.randn(5, 3)
+    >>> mat = diag(val)
+    >>> mat.shape
+    (5, 5)
+    >>> mat.nnz
+    5
+    """
+    # Mufei: this may not be needed if DiagMatrix is simple enough
+    return DiagMatrix(val, shape)
+
+def identity(shape: Tuple[int, int],
+             d: Optional[int] = None,
+             dtype: Optional[torch.dtype] = None,
+             device: Optional[torch.device] = None) -> DiagMatrix:
+    """Create a diagonal matrix with ones on the diagonal and zeros elsewhere
+
+    Parameters
+    ----------
+    shape : tuple[int, int]
+        Shape of the matrix.
+    d : int, optional
+        If None, the diagonal entries will be scaler 1. Otherwise, the diagonal
+        entries will be a 1-valued tensor of shape (d).
+    dtype : torch.dtype, optional
+        The data type of the matrix
+    device : torch.device, optional
+        The device of the matrix
+
+    Returns
+    -------
+    DiagMatrix
+        Diagonal matrix
+
+    Examples
+    --------
+
+    Case1: 3-by-3 matrix with scaler diagonal values
+
+    [[1, 0, 0],
+     [0, 1, 0],
+     [0, 0, 1]]
+
+    >>> mat = identity(shape=(3, 3))
+    >>> mat.val
+    tensor([1., 1., 1.])
+    >>> mat.shape
+    (3, 3)
+    >>> mat.nnz
+    3
+
+    Case2: 3-by-5 matrix with scaler diagonal values
+
+    [[1, 0, 0, 0, 0],
+     [0, 1, 0, 0, 0],
+     [0, 0, 1, 0, 0]]
```

Comment:
I believe identity matrix is always square.

Reply:
Not necessarily, e.g., when adding self-loops for a `DGLBlock`.

Reply:
> e.g., when adding self-loops for a `DGLBlock`.

I think it's ok to have such support but want to point out that maybe the easiest way to add selfloop to `DGLBLock` is using operation like `A[I, I] = val`.

Reply:
Sure.